REGR: GH50127 fixes error for converting to / creating Index with dtype `S3`

Daquisu opened the pull request:
- [x] closes #50127
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions. (Does not apply)
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -501,6 +501,11 @@ def __new__(
                 # should not be coerced
                 data = com.asarray_tuplesafe(data, dtype=_dtype_obj)
+                # # GH#50127 we must update the `dtype` when we have the numpy
+                # # type `S` to `_dtype_to_subclass`, because it would raise a
```
i dont think we need the double-"#"

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -532,10 +537,13 @@ def __new__(
+            elif dtype and dtype.kind == "S":
+                # GH#50127 we update data to a np.array with the correct dtype.
+                data = astype_array(np.array(data, copy=copy), dtype=dtype, copy=copy)
+                return Index(data, dtype=dtype, copy=copy, name=name)
```
is there a viable way to avoid recursing here?

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -532,10 +537,13 @@ def __new__(
                 data = com.asarray_tuplesafe(data, dtype=_dtype_obj)
+            elif dtype and dtype.kind == "S":
+                # GH#50127 we update data to a np.array with the correct dtype.
+                data = astype_array(np.array(data, copy=copy), dtype=dtype, copy=copy)
```
why not just set `dtype = _dtype_obj` here like on L508?

jbrockmendel replied:
can we avoid passing copy=copy twice here?

MarcoGorelli commented:
moving off the 2.0 milestone as this is a regression from 1.5

jbrockmendel commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -1399,6 +1399,7 @@ Other
 - Bug in incorrectly accepting dtype strings containing "[pyarrow]" more than once (:issue:`51548`)
 - Bug in :meth:`Series.searchsorted` inconsistent behavior when accepting :class:`DataFrame` as parameter ``value`` (:issue:`49620`)
 - Bug in :func:`array` failing to raise on :class:`DataFrame` inputs (:issue:`51167`)
+- Bug in :class:`Index` when created or converted to the numpy dtype S3 would raise ``NotImplementedError`` (:issue:`50127`)
```
"when created or converted" -> "where creating or converting"?

datapythonista commented:
Sorry, my bad, the next release should be 2.0.1, changed the milestone again.

jbrockmendel commented on doc/source/whatsnew/v2.0.1.rst:
```diff
@@ -32,7 +32,7 @@ Bug fixes
 Other
 ~~~~~
--
+- Bug in :class:`Index` where creating or converting to the numpy dtype S3 would raise ``NotImplementedError`` (:issue:`50127`)
```
instead of specifically S3 should this be numpy string dtypes more generally?

Daquisu replied:
Done.

jbrockmendel commented on pandas/core/indexes/base.py:
```diff
@@ -543,10 +548,13 @@ def __new__(
                 data = com.asarray_tuplesafe(data, dtype=_dtype_obj)
+            elif dtype and dtype.kind == "S":
+                # GH#50127 we update data to a np.array with the correct dtype.
+                data = np.array(data, dtype=object, copy=copy)
```
after doing this can we set copy=False to ensure no subsequent copies

Daquisu replied:
Done.

Daquisu commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -1137,3 +1138,15 @@ def test_multi_column_dtype_assignment():
+
+
+def test_numpy_s3_dtype_on_index():
+    # GH #50127
```
Done.

Daquisu commented:
` Ubuntu / Numpy Dev (pull_request)` is failing on all recent PRs, so I guess it is not related to the changes done here.

topper-123 committed:
REF: Decouple Series.apply from Series.agg (#53400)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53536)

Examples

liang3zy22 committed:
TST: Add test for boxable categories GH#21658 (#53524)

Signed-off-by: Liang Yan <ckgppl_yan@sina.cn>

mroeschke committed:
CI: Bump builds to 3.11, update numpy nightly wheels location (#53508)

* CI: Bump builds to 3.11, update numpy nightly wheels location

* Move arm to 3.10 instead

* Change scipy window api

* plural

* Use stable sort?

jbrockmendel committed:
DEPR: use_inf_as_na (#53494)

* DEPR: use_inf_as_na

* troubleshoot doc

* troubleshoot docbuild

* troubleshoot doc

* troubleshoot docs

Charlie-XIAO committed:
BUG: `groupby` with column selection not returning tuple when grouping by list of a single element (#53517)

mroeschke committed:
TST: Remove inheritance from pandas/tests/indexes  (#53544)

* Refactor test/indexes

* Address failures

lukemanley committed:
BUG: Series.str.split(expand=True) for ArrowDtype(pa.string()) (#53532)

* BUG: Series.str.split(expand=True) for ArrowDtype(pa.string())

* whatsnew

* min versions

* ensure ArrowExtensionArray

lithomas1 committed:
TST: Clean up finalize tests (#53543)

lithomas1 committed:
BUG: Fix metadata propagation in squeeze and describe (#53538)

* BUG: Fix metadata propagation in squeeze and describe

* remove unnecessary param

topper-123 committed:
DEPR: make Series.agg aggregate when possible (#53325)

* BUG: make Series.agg aggregate when possible

* fix doc build

* deprecate instead of treating as a bug

* CLN: Apply.agg_list_like

* some cleanups

* REF/CLN: func in core.apply (#53437)

* REF/CLN: func in core.apply

* Remove type-hint

* REF: Decouple Series.apply from Series.agg (#53400)

* update test

* fix issues

* fix issues

* fix issues

---------

Co-authored-by: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

datapythonista committed:
DOC: Update release process (#53435)

* DOC: Update release process

* Add exact symlinks instructions

DeaMariaLeon committed:
DOC: Fixing EX01 - Added more examples (#53540)

* Added examples groupby

* Removed decorator & added 'See Also' inside docstring

* Added @Substitution instead of repeating 'see_also'

liang3zy22 committed:
TST: Add test for timedelta of fraction GH#9048 (#53552)

Signed-off-by: Liang Yan <ckgppl_yan@sina.cn>

liang3zy22 committed:
TST: Add test for complex categories GH#16399 (#53551)

Signed-off-by: Liang Yan <ckgppl_yan@sina.cn>

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53554)

* Example for count

* Added examples

jbrockmendel committed:
DEPR: fillna method kwd (#53496)

* DEPR: fillna method kwd

* deprecate Resampler.fillna entirely

* suppress doctest warnings

* typo fixup

* doctest

* mypy fixup

* lint fixup

lithomas1 committed:
ENH: Numba groupby support multiple labels (#53556)

* ENH: Numba groupby support multiple labels

* update regex

okjustin committed:
Update 02_read_write.rst (#53559)

Grammatical error corrected.

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53561)

* testing example in template

* Added max example

* Added groupby examples

* Updated code_checks

* Corrected alignment

mroeschke committed:
CI/DEPS: Add xfail(strict=False) to related unstable sorting changes in Numpy 1.25 (#53548)

* DEBUG: npdev build

* Address tests where sorting changed

* Adjust more tests

* Undo everything, even nanargsort

* xfail the relevant tests

* Add xfail to test_sort_column_level_and_index_label

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53564)

* Example for pct_change

* Added examples for groupby sem, shift, size

* Updated code_checks

* Corrected error on groupby size

jarrodmillman committed:
Upload nightlies to new location (#53341)

phofl committed:
CLN: Cleanup after CoW setitem PRs (#53142)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53573)

Added examples

mroeschke committed:
TST: Use more pytest fixtures (#53567)

* TST: Use more fixtures

* Use more fixtures in test_indexing_slow

* Move addition compression_to_extension

* Use more fixture in sparse test_indexing

* fixturize libsparse

mroeschke committed:
TST: Use fixtures instead of TestPlotBase (#53550)

* TST: Use fixtures instead of TestPlotBase

* move importskip to fixture

natmokval committed:
DOC: create a table for period aliases (#53530)

* DOC: table for period aliases

* correct the to_period docstrings

* change quarter/year end to quarterly/yearly

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53575)

Added examples

Charlie-XIAO committed:
DEPR `fill_method` and `limit` keywords in `pct_change` (#53520)

* DEPR fill_method and limit keywords in DataFrame/Series pct_change

* DEPR fill_method and limit keywords in GroupBy pct_change

* changelog added

* DataFrame/Series pct_change only deprecate default

* typo

* GroupBy pct_change only deprecate default

* changelod updated correspondingly

* reverting

* Revert "reverting"

This reverts commit 2cb449bab9ede4440d01d8158c9d6b44ecc4632e.

* resolved conversation: also warn if nan and not specified

* modify msgs since fillna method deprecated; tests updated

* changelog updated

* docstring use ffill instead of fillna

mroeschke committed:
TST: Use monkeypatch instead of custom func for env variables (#53563)

* TST: Use monkeypatch instead of custom func for env variables

* Make scopes match

mroeschke committed:
CI: Ensure nightly wheels are uploaded instead of just sdist (#53578)

behrenhoff committed:
TYP: Fix typing issues, mainly Something => IndexLabel (#53469)

TYP: Fix typing issues, Label => IndexLabel

kyunghei committed:
Corrected minor typos in the paragraph below 'Formatting the Display' (#53598)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53619)

* ctime docstring

* DOC Timestamp.as_unit and ctime

lithomas1 committed:
CI: Fix the deprecation bot (#53593)

* CI: Fix the deprecation tracking bot

* Update deprecation-tracking-bot.yml

* Update deprecation-tracking-bot.yml

* Updates

* Update deprecation-tracking-bot.yml

dependabot[bot] committed:
Bump pypa/cibuildwheel from 2.13.0 to 2.13.1 (#53614)

Bumps [pypa/cibuildwheel](https://github.com/pypa/cibuildwheel) from 2.13.0 to 2.13.1.
- [Release notes](https://github.com/pypa/cibuildwheel/releases)
- [Changelog](https://github.com/pypa/cibuildwheel/blob/main/docs/changelog.md)
- [Commits](https://github.com/pypa/cibuildwheel/compare/v2.13.0...v2.13.1)

---
updated-dependencies:
- dependency-name: pypa/cibuildwheel
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mroeschke committed:
TST: Mark numba & threading tests as single cpu (#53608)

* TST: Mark numba & threading tests as sinngle cpu

* Update pandas/tests/window/test_online.py

ABizzinotto committed:
DOC: pd.Period and pd.period_range should document that they accept datetime, date and pd.Timestamp (#53611)

* Add additional valid types to 'value'

* Use pandas.Timestamp instead of pd.Timestamp

tpaxman committed:
DOC: Fixed inconsistencies in pandas.DataFrame.pivot_table docstring parameter descriptions (#53605)

DOC: Inconsistencies in pandas.DataFrame.pivot_table parameter descriptions (#53351)

jbrockmendel committed:
ENH: Implement PandasArray, DTA, TDA interpolate (#53594)

* ENH: implement PandasArray.interpolate

* Implement DTA/TDA.interpolate

lukemanley committed:
PERF: Series.str.split(expand=True) for pyarrow-backed strings (#53585)

* PERF: Series.str.split(expand=True) for pyarrow-backed

* gh ref

sweisss committed:
DOC: Multi-conditional examples added to .loc docstring (#53572)

* Add multi-conditional examples in loc docstring

* Update boolean operators in example descriptions

* Update note on advanced indexing

* Add point to wrap conditionals in parens.

jbrockmendel committed:
DEPR: pd.value_counts (#53493)

* DEPR: pd.value_counts

* update docs

* suppress warning

* suppress warning

* suppress warning

* suppress warning

* pyright ignore

* pylint ignore

* warn inside value_counts

jbrockmendel committed:
REF: move interpolate validation to core.missing (#53580)

* REF: move interpolate validation to core.missing

* lint fixup

lithomas1 committed:
ENH: Groupby.transform support string input with engine=numba (#53579)

* ENH: Groupby.transform support string input with engine=numba

* remove xfail

jbrockmendel committed:
DEPR: NDFrame.interpolate with ffill/bfill methods (#53607)

* DEPR: NDFrame.interpolate with ffill/bfill methods

* update doctest

jbrockmendel committed:
BUG: clean_fill_method failing to raise (#53620)

* BUG: clean_fill_method failing to raise

* mypy fixup

* Update doc/source/whatsnew/v2.1.0.rst

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

* sort note

---------

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

datapythonista committed:
PDEP-9: Allow third-party projects to register pandas connectors with a standard API (#51799)

liang3zy22 committed:
TST: Add test for apply cast types GH#9506 (#53591)

Signed-off-by: Liang Yan <ckgppl_yan@sina.cn>

rhshadrach committed:
BUG: DataFrame.stack with sort=True and unsorted MultiIndex levels (#53637)

* BUG: DataFrame.stack with sort=True and unsorted MultiIndex levels

* revert ignoring warnings

lukemanley committed:
ENH: Series.str.join for ArrowDtype(pa.string()) (#53646)

* Series.str.join to support ArrowDtype(pa.string())

* gh ref

jbrockmendel committed:
DEPR: interpolate with object dtype (#53638)

jbrockmendel committed:
BUG: interpolate with complex dtype (#53635)

* BUG: interpolate with complex dtype

* GH ref

ssharp0 committed:
DOC: Added links to GitHub tutorials/resources for Forking Workflow in contributing.rst (#53628)

* Added links to GitHub tutorials/resources for forking workflows in contributing.rst

* updated with shortened link and minor reformatting

mroeschke committed:
TST: Refactor slow test (#53610)

mroeschke committed:
BUG: RangeIndex.union(sort=True) with another RangeIndex (#53495)

lukemanley committed:
ENH: Series.explode to support pyarrow-backed list types (#53602)

* ENH: Series.explode to support pyarrow-backed list types

* gh refs

* update test

mroeschke committed:
CI: Build pandas even if doctests fail (#53657)

Charlie-XIAO committed:
BUG: groupby sum turning `inf+inf` and `(-inf)+(-inf)` into `nan` (#53623)

jbrockmendel committed:
DEPR: method, limit in NDFrame.replace (#53492)

* DEPR: method, limit in NDFrame.replace

* update test, docs

* suppress doctest warning

* doctests

lukemanley committed:
PERF: Series.str.get_dummies for ArrowDtype(pa.string()) (#53655)

* PERF: Series.str.get_dummies for ArrowDtype(pa.string())

* whatsnew

* typing

jbrockmendel committed:
TYP: core.missing (#53625)

lithomas1 committed:
CI: Attempt to fix wheel builds (#53670)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53647)

* SeriesGroupBy.fillna example added

* Added examples

* Corrected failing test for timedelta.total_seconds

* Corrected fillna example

mroeschke committed:
CI/TST: Mark test_to_read_gcs as single_cpu (#53677)

lithomas1 committed:
BUG/CoW: is_range_indexer can't handle very large arrays (#53672)

* BUG: is_range_indexer can't handle very large arrays

* fix test on 32-bit

lithomas1 committed:
ENH: Allow numba aggregations to return non-float64 results (#53444)

* ENH: non float64 result support in numba groupby

* refactor & simplify

* fix CI

* maybe green?

* skip unsupported ops in other bench as well

* updates from code review

* remove commented code

* update whatsnew

* debug benchmarks

* Skip min/max benchmarks

lithomas1 committed:
DOC: Move Whatsnew for CoW fix (#53690)

* DOC: Move Whatsnew for CoW fix

* 2.0.3

lithomas1 committed:
CI: Adjust tests for release of numpy 1.25 (#53715)

natmokval committed:
DEPR: deprecate obj argument in GroupBy.get_group (#53571)

jorisvandenbossche committed:
Use _values_for_factorize by default for hashing ExtensionArrays (#53475)

MarcoGorelli committed:
CI: ignore experimental warnings from numba (#53726)

MarcoGorelli committed:
CI: fix `pytest scripts` (#53727)

jorisvandenbossche committed:
TST: add test for reindexing rows with matching index uses shallow copy with CoW (#53723)

phofl committed:
BUG: IntervalIndex.get_indexer raising for read only array (#53703)

rhshadrach committed:
ERR: Shorten traceback in groupby _cython_agg_general (#52992)

phofl committed:
CoW: Add lazy copy mechanism to DataFrame constructor for dict of Index (#52947)

MarcoGorelli committed:
CI: fix numba typing (#53730)

* fix numba typing

* type ignore

---------

Co-authored-by: Patrick Hoefler <61934744+phofl@users.noreply.github.com>

natmokval committed:
DEPR: deprecate unit parameters ’T’, 't', 'L', and 'l' (#53557)

* BUG: add ’T’ to UnitChoices to fix incorrect unit validation in to_timedelta()

* deprecate units T, t, L, l and fix tests

* raise FutureWarning instead of ValueError

* correct the to_timedelta definition and add a test for the warning

---------

Co-authored-by: Marco Edward Gorelli <marcogorelli@protonmail.com>

phofl committed:
CoW: Return read-only array in Index.values (#53704)

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53689)

* Added Timestamp examples

* Added test and corrected msg

* Correcting error on isoweekday

* Added requested corrections

* added import datetime

* Removed isoweekday raise error, corrected time and timetz

mroeschke committed:
TST: Use more pytest fixtures (#53679)

* TST: Use more pytest fixtures

* Fix sql fixture

* More test fixtures

Charlie-XIAO committed:
BUG: resampling empty series loses time zone from dtype (#53736)

* BUG: resampling empty series loses time zone from dtype

* changelog added

phofl committed:
BUG: concat coercing arrow to object with null type (#53702)

* BUG: concat coercing arrow to object with null type

* BUG: concat coercing arrow to object with null type

lukemanley committed:
PERF: concatenation of MultiIndexed objects (MultiIndex.append) (#53697)

* improve perf of MultiIndex.append

* fix test

* style

liang3zy22 committed:
TST: Add test for duplcated columns and usecols GH#11823 (#53683)

Signed-off-by: Liang Yan <ckgppl_yan@sina.cn>

Charlie-XIAO committed:
BUG: series with complex nan (#53682)

* avoid float(complex('nan'))

* update

* added test

* modified tests; added changelog

* changelog modified

lukemanley committed:
BUG: astype('category') on dataframe backed by non-writeable arrays raises ValueError (#53678)

Jython1415 committed:
TST: xfail `test_rolling_var_numerical_issues` on Mac (#53661)

modified xfail test_rolling_var_numerical_issues
removed exclusion of mac from xfail

mcgeestocks committed:
BUG: Adds missing raises for numpy.timedelta64[M/Y] in pandas.Timedelta (#53421)

* adds raise for numpy Y and M conversion in Timedelta

* fix failing tests

* update raise lang

* alternative approach

* add missing periods

* switch to cimport

* add cython directive

* switch to inline cython directive

* remove unsupported td operation

* adds whatnew update

* remove whitespace

* adds test

rmhowe425 committed:
DEPR: Deprecate literal json string input to read_json (#53409)

* Adding logic to throw a deprecation warning when a literal json string is passed to read_json

* Adding logic to throw a deprecation warning when a literal json string is passed to read_json

* Updating documentation and adding PR num to unit test

* Adding a deprecation warning to the user guide

* Updating unit tests to check for FutureWarning

* Fixing unit tests

* Fixing unit tests

* Fixing unit tests

* Fixing unit tests

* Fixing documentation errors in PR feedback

* Fixing documentation errors in PR feedback

* Updating unit tests to use StringIO rather than catch FutureWarning

* Finishing updating unit tests to use StringIO rather than catch FutureWarning

* Fixing indendation errors in unit tests. Moved one unit test to another file.

* Updating unit test name

* Adding additional checks to unit tests

* Fixing unit tests

* Fixing unit tests

* Updating whatsnew documentation per reviewer recommendations.

* Fixing failing code tests

* Fixing failing code tests

* Adding import to doc string example

* Fixing documentation formatting error

* Fixing documentation formatting error

* Fixing documentation error after fixing merge conflict

* Fixing formatting errors in whatsnew file

* Updating formatting errors in documentation

* Updating formatting errors in documentation

mroeschke committed:
DOC: Add to_sql example of conflict with `method` parameter (#53264)

* DOC: Add to_sql example of conflict do nothing

* Use fixture to skip

* Add example of conflict update

mroeschke committed:
BUG: reindex with expansion and non-nanosecond dtype (#53505)

* BUG: reindex with expansion and non-nanosecond dtype

* Restrict to timelike types

* Check earlier

* handle NA

* handle NA

* Better check

mroeschke committed:
BUG: convert_dtype(dtype_backend=nullable_numpy) with ArrowDtype (#53651)

mroeschke committed:
BUG: Indexing a timestamp ArrowDtype Index (#53652)

* BUG: Indexing a timestamp ArrowDtype Index

* more explicit check

jbrockmendel committed:
TYP: reshape.merge (#53752)

mroeschke committed:
TST: Use more pytest fixtures (#53750)

* CI: Build pandas even if doctests fail

* BUG: groupby sum turning `inf+inf` and `(-inf)+(-inf)` into `nan` (#53623)

* DEPR: method, limit in NDFrame.replace (#53492)

* DEPR: method, limit in NDFrame.replace

* update test, docs

* suppress doctest warning

* doctests

* PERF: Series.str.get_dummies for ArrowDtype(pa.string()) (#53655)

* PERF: Series.str.get_dummies for ArrowDtype(pa.string())

* whatsnew

* typing

* TYP: core.missing (#53625)

* CI: Attempt to fix wheel builds (#53670)

* DOC: Fixing EX01 - Added examples (#53647)

* SeriesGroupBy.fillna example added

* Added examples

* Corrected failing test for timedelta.total_seconds

* Corrected fillna example

* CI/TST: Mark test_to_read_gcs as single_cpu (#53677)

* BUG/CoW: is_range_indexer can't handle very large arrays (#53672)

* BUG: is_range_indexer can't handle very large arrays

* fix test on 32-bit

* TST: Use more pytest fixtures

---------

Co-authored-by: Yao Xiao <108576690+Charlie-XIAO@users.noreply.github.com>
Co-authored-by: jbrockmendel <jbrockmendel@gmail.com>
Co-authored-by: Luke Manley <lukemanley@gmail.com>
Co-authored-by: Thomas Li <47963215+lithomas1@users.noreply.github.com>
Co-authored-by: Dea María Léon <deamarialeon@gmail.com>

mroeschke committed:
BUG: DataFrame construction with dictionary ArrowDtype columns (#53654)

* BUG: DataFrame construction with dictionary ArrowDtype columns

* Add tests for get_indexer

* Windows

natmokval committed:
DOC: point out that only period aliases are valid for the method asfreq in period.pyx (#53751)

DOC: point out that only period aliases are valid for Period method asfreq

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53741)

* wip

* Examples for tzname, utcoffset, utctimetuple, weekday, strptime

rsm-23 committed:
DOC: operator wrapper names do not match math operators (#53765)

* Updated DataFrame.div docstrings

* Tagged changes to v2.0.3

* Removed whatsnew comment

* Update v2.0.3.rst

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53725)

* Examples Timestamp.time, timetuple, timetz, to_datetime64, toordinal

* Added tests and updated code_checks.sh

* Corrected time and timetz

* Corrected Timestamp.time and timetz

rmhowe425 committed:
DEPR: Deprecate DataFrame.last and Series.last (#53710)

* Adding implementation for deprecation and entry in whatsnew file

* Updating unit tests to account for deprecating of series.last() and DataFrame.last()

* Added deprecation message in doc string

* Adding PR number to new unit test

* Removing duplicate "Parameters" docstring header

* Adding doctest skip for call to last()

phofl committed:
BUG / CoW: Series.transform not respecting CoW (#53747)

rhshadrach committed:
Revert "BUG: DataFrame.stack with sort=True and unsorted MultiIndex levels" (#53760)

Revert "BUG: DataFrame.stack with sort=True and unsorted MultiIndex levels (#53637)"

This reverts commit 5edc2cc570f565326cb9a762d3f19a19f6bab425.

fangchenli committed:
TST: refactor data path for xml tests (#53766)

* TST: refactor data path for xml tests

* fix style

* fix typo

mroeschke committed:
TST: Reduce memory pressure of plotting tests (#53660)

* TST: Reduce memory pressure of plotting tests

* Trigger ci

* Remove gc call

MarcoGorelli committed:
DOC note pytest bump (#53768)

* note pytest bump

* update install.rst

mroeschke committed:
COMPAT: Remove np.compat (#53774)

COMPAT: Remove np.comat

rahulsiloniya committed:
Added suggested new line to fix doc code example (#53775)

* Added suggested new line to fix doc code example

* Removed the newline

* Done

mroeschke committed:
TST: Make test_complibs deterministic (#53754)

* TST: Make test_complibs deterministic

* Make sure files are unique?

* Try unique key

* Just xfail test

* Check what is taking long

* Remove -v

* Filter warning

mroeschke committed:
TST: Refactor some slow tests (#53784)

* Cleanup single used method

* Clean plotting test

* Improve test_series_groupby_nunique

* Address more slow tests

* Undo changes

jbrockmendel committed:
TYP: reshape.merge (#53780)

fangchenli committed:
TYP: annotate testing decorators with pytest.MarkDecorator (#53794)

TYP: type pytest.MarkDecorator

fangchenli committed:
TST/CLN: use fixture for data path in all xml tests (#53790)

TST/CLN: use fixture path for all xml tests

jbrockmendel committed:
REF: remove unused merge args (#53789)

phofl committed:
BUG: combine_first ignoring others columns if other is empty (#53792)

* BUG: combine_first ignoring others columns if other is empty

* Fix

jbrockmendel committed:
PERF: concat in no-reindexing case (#53772)

Charlie-XIAO committed:
BUG: bad display for complex series with nan (#53764)

* BUG: bad display for complex series with nan

* added comments

* added more test cases

jbrockmendel committed:
CLN: assorted (#53742)

* CLN: assorted

* revert pivot edits

* revert np.float128 check

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53796)

* examples TimedeltaArray, Period.asfreq

* Examples for categoricals

* Exampl Categorical.codes and .__array__, edited code_checks

* Corrected TimedeltaArray and wording to categorical

* remove multiline

---------

Co-authored-by: MarcoGorelli <33491632+MarcoGorelli@users.noreply.github.com>

computerscienceiscool committed:
Update governance.md (#53814)

* Update governance.md

Misspelling of Subcommittee

* Update web/pandas/about/governance.md

---------

Co-authored-by: Marco Edward Gorelli <marcogorelli@protonmail.com>

rsm-23 committed:
BUG: inconsistent behavior of DateOffset #47953 (#53681)

* BUG: Fixed inconsistent multiplication #47953

* Fixed release note

* Fixed attribute name

* Changed offset apply logic

* Addressed #46877 re-occurence

* Removed dulicate code

* Addressed comments

* Added unit test cases

* Added mistakenly removed comment

ABizzinotto committed:
DOC: pd.Period and pd.period_range should document that they accept datetime, date and pd.Timestamp (#53632)

* Update docstrings for pandas.Period and period_range

* Fix Period docstring comment

* Fix Period value docstring

MarcoGorelli committed:
CLN: PDEP6 precusor (#53815)

set dtype explicitly

tpaxman committed:
DOC: Add code-formatting and cross-reference links to `read_csv` docstring (#53735)

* apply appropriate code-formatting to `read_csv` docstring using backticks

* escape curly braces for `dict` examples

* fix formatting under `delim_whitespace` so that tab character appears properly

* fix formatting issue in read_csv description for dtype_backend

* test linking for :class:`~pandas.DataFrame`

* add spinx references to pandas classes and functions

* remove code formatting from initial parameter description lines in read_csv

* replaced "default None" with "optional" in read_csv docstring

* add additional code formatting changes to read_csv after review

* shorten docstring lines to 88 characters max

* fix formatting causing docstring error

---------

Co-authored-by: tpaxman <tpaxman@DESKTOP-N3NLG1E.localdomain>

mroeschke committed:
TST: Refactor more slow tests (#53800)

* TST: Refactor more slow test

* more tests

* define less fixtures

mroeschke committed:
TST: Refactor s3 resource (#53803)

* TST: Refactor s3 resource

* Fix api usage

* bot3 instead of cli

* refactor call

* Make cleanup and bucket names more unique

* Use external error raised

jbrockmendel committed:
REF: simplify merge code (#53808)

* REF: simplify merge code

* typing

liang3zy22 committed:
TST: Add test for loc expansion GH13829. (#53812)

Signed-off-by: Liang Yan <ckgppl_yan@sina.cn>

jbrockmendel committed:
REF: de-duplicate Block.diff, remove axis kwd (#53807)

* REF: remove axis keyword from internal diff methods

* De-duplicate diff

lukemanley committed:
PERF: pandas.core.sorting.compress_group_index for already sorted values (#53806)

* PERF: pandas.core.sorting.compress_group_index for already sorted values

* whatsnew

rhshadrach committed:
BUG: DataFrame.stack sorts columns (#53787)

* BUG: DataFrame.stack with sort=True and unsorted MultiIndex levels

* revert ignoring warnings

* BUG: DataFrame.stack sorting columns

* whatsnew

* Docstring fixup

* Merge cleanup

ssche committed:
Fixed #52234 (#53698)

* Fixed #52234

* Copied fix from https://github.com/ltauro/pandas/tree/nan-not-in-index and applied review comments

* Update doc/source/whatsnew/v2.1.0.rst

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

---------

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

mroeschke committed:
TST/CI: Skipif test_complibs (#53819)

jbrockmendel committed:
REF: separate out cross-merge, make less stateful (#53810)

jbrockmendel committed:
REF: implement NDFrame._from_mgr (#52132)

jbrockmendel committed:
REF: implement Manager.pad_or_backfill (#53822)

jbrockmendel committed:
PERF: avoid unnecessary casting in merge (#53771)

phofl committed:
TST / CoW: Add test for mask (#53745)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53818)

* Examples api.types, Sparse, TZDtype

* Edited code_checks

* Changed tz to zoneinfo

* Added ZoneInfo for tz param

yuanx749 committed:
BUG: Fix string formatting (#53855)

* fixed string formatting bug

* updating doc

* add tests

* change whatsnew

* Update doc/source/whatsnew/v2.0.1.rst

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

* Update v2.0.1.rst

* Update v2.0.1.rst

* change whatsnew

* Update pandas/tests/io/formats/test_format.py

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

* Fix bug with smaller change

---------

Co-authored-by: Rithik Reddy <rithik.g.reddy@vanderbilt.edu>
Co-authored-by: reddyrg1 <59897313+reddyrg1@users.noreply.github.com>
Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

tpaxman committed:
DOC: parameterize "See Also" content in `read_table` and `read_csv` so that only the other function appears (#53848)

* parameterize the "see also" function for read_csv and read_table docstrings to always show the other

* fix long line formatting issue

mroeschke committed:
TST: Refactor more slow tests (#53820)

tpaxman committed:
DOC: update `read_clipboard` docstring to have proper formatting and all required sections. (#53838)

* add code formatting and reference links to read_clipboard docstring

* add extended summary, see-also, and examples sections to ensure docstring is fully validated

* hide "pandas" from each :func: and :class: reference

* fix typos

tpaxman committed:
DOC: Add description to `read_csv` chunksize parameter regarding number of lines in chunk (#53836)

add description to read_csv chunksize parameter

tpaxman committed:
DOC: fix inconsistencies in `read_csv` docstring type descriptions (#53834)

* update type descriptions for read_csv parameters

* make read_csv parameter descriptions consistent with valid type options

* correct minor docstring formatting

Charlie-XIAO committed:
BUG: complex Series/DataFrame display all complex nans as `nan+0j` (#53844)

* BUG: complex Series/DataFrame display all complex nans as nan+0j

* resolve mypy via type: ignore[arg-type]

jbrockmendel committed:
BUG: DataFrame.shift(axis=1) with EADtype (#53832)

* BUG: DataFrame.shift(axis=1) with EADtype

* GH refs, final

jbrockmendel committed:
REF: implement PandasArray.pad_or_backfill (#53827)

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

NMAC427 committed:
BUG: to_sql fails for pyarrow date dtype (#53854) (#53856)

* BUG: to_sql fails for pyarrow date dtype (#53854)

* use ArrowExtensionArray.to_numpy instead of ArrowExtensionArray.astype

mroeschke committed:
TYP: Fix pad_or_backfill limit_area type (#53866)

mroeschke committed:
TST: Clean plotting/test_series (#53865)

* clean test_series

* Change to importskip

mroeschke committed:
TST: Clean plotting/test_misc.py (#53864)

clean plotting test_misc

mroeschke committed:
TST: Simplify more slow tests (#53862)

* Clean more slow tests

* Closer scope

jbrockmendel committed:
BUG: merge with mismatched index dtypes failing to raise (#53870)

* BUG: merge with mismatched index dtypes failing to raise

* gh ref

jbrockmendel committed:
REF: share methods AM/BM (#53859)

natmokval committed:
TST: increase coverage of asfreq (#53861)

* TST: increase coverage of asfreq

* remove redundant argument

rsm-23 committed:
DOC: Updated docstring for Period class (#53879)

Updated docstring for Period class

mcgeestocks committed:
BUG: fixes weekday for dates before 1752 (#53795)

* switch to gauss algo

* precommit cleanup

* missed formatting and whatsnew

* add hypothesis and negative test

* update parametric test

* requested changes

* move whatsnew to 2.0.3

* post-merge fixup

* note source of implementation, simplify, include negative dates in parametric test

---------

Co-authored-by: MarcoGorelli <33491632+MarcoGorelli@users.noreply.github.com>
Co-authored-by: Marco Edward Gorelli <marcogorelli@protonmail.com>

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53883)

* Examples to_html, put, append

* Examples to_html, put, append, get, select, info, keys

* Correct select & keys

mroeschke committed:
TST: Use pytest-localserver instead of making network connections (#53828)

* Use pytest-localserver instead of making network connections

* Fix test, remove network function

* remove network from init

* Ignore disutils from datareader, s3so

* specify encoding

* Specify encoding

* Clarify contributing doc

mroeschke committed:
TST: Clean plotting/frame/test_frame.py (#53873)

* clean test frame

* finish off tests

jbrockmendel committed:
CLN: unused args in pad_or_backfill (#53872)

* CLN: unused args in pad_or_backfill

* restore limit_direction validation

Charlie-XIAO committed:
CLN: use `lib.no_default` instead of `lib.NoDefault` in `.pivot` (#53877)

* CLN typo lib.NoDefault -> lib.no_default

* also typo in tests

Pawel-Kranzberg committed:
DOC Remove incorrect line from dropna docs (#53876)

Update frame.py - remove incorrect line from dropna docs

Remove incorrect line "Pass tuple or list to drop on multiple axes. " from dropna, as the ability to drop on multiple axes has been depreciated.

jbrockmendel committed:
REF: simplify _from_mgr constructors, use in more places (#53871)

jbrockmendel committed:
REF: helpers to de-duplicate CoW checks (#53882)

* REF: refactor out Block helpers for CoW

* REF: de-duplicate helpers

* De-duplicate helper, revert _replace_coerce edits

jbrockmendel committed:
REF: remove axis keyword from Manager/Block.shift (#53845)

* REF: remove axis keyword from Manager/Block.shift

* mypy fixup

* mypy fixup

rmhowe425 committed:
DEPR: Remove bytes input for read_excel (#53830)

* Adding deprecation logic and updating documentation.

* Adding PR number to new unit test

* Updating read_excel() documentation per reviewer recommendations.

* Updating read_excel() documentation per reviewer recommendations.

* Updating read_excel() documentation per reviewer recommendations.

* Updating implementation per reviewer recommendations.

jbrockmendel committed:
REF: de-duplicate methods, remove unused kwd (#53886)

* REF: dedup Manager method, remove Manager.quantile axis kwd

* remove unused kwd

jbrockmendel committed:
REF: de-duplicate _gotitem helper (#53874)

* REF: de-duplicate _gotitem helper

* mypy fixup

mroeschke committed:
CI: Fix pytest scripts and doc failure (#53888)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53889)

Examples HDFStore.groups, walk

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53880)

* read_clipboard

* Examples read_clipboard, ExcelFile, ExcelFile.parse

* skip test for read_clipboard

* Added to ExcelFile example

natmokval committed:
TST: add test with freq=2M for period_array from datetime64 (#53890)

Charlie-XIAO committed:
DOC: fix asv test results link in contributors doc (#53902)

phofl committed:
BUG: df constructor not copying ea backed series (#53744)

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

jbrockmendel committed:
REF: de-duplicate CoW helpers (#53899)

mroeschke committed:
TST: Clean plotting/frame/test_frame.color.py (#53894)

rhshadrach committed:
BUG: DataFrame.stack sometimes sorting the resulting index (#53825)

* BUG: DataFrame.stack with sort=True and unsorted MultiIndex levels

* revert ignoring warnings

* BUG: DataFrame.stack sorting columns

* whatsnew

* Docstring fixup

* Merge cleanup

* WIP

* BUG: DataFrame.stack sometimes sorting the resulting index

* mypy fixups

* Remove sort argument from DataFrame.stack

rmhowe425 committed:
DEPR: Remove literal string input for read_html (#53805)

* Updating documentation and adding deprecation logic for read_html.

* Fixing formatting errors

* Fixing documentation errors

* Updating deprecation logic and documentation per reviewer recommendations.

* Updating implementation per reviewer recommendations.

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53906)

* Examples Resampler.__iter__, groups, indices, get_group, ffill

* Improved explanation for ffill

lithomas1 committed:
DOC: Update release date for 2.0.3 (#53897)

mroeschke committed:
TST: Refactor slow tests (#53891)

* Address more slow tests

* Parameterize slow test

* Reduce data size in multi_thread

* Use constant data for test_int64_overflow_groupby_large_df_shuffled

jbrockmendel committed:
ENH: EA.interpolate (#53659)

* ENH: EA.interpolate

* update code check

* update docstring to appease code check

* inplace->copy

* whatsnew

mroeschke committed:
TST: Clean tests/plotting/test_frame.py misc (#53914)

jbrockmendel committed:
DEPR: downcast keyword (#53671)

mroeschke committed:
CI: Don't rebuild pandas before running tests (#53919)

mroeschke committed:
TST: Clean plotting tests (#53917)

* TST: Clean plotting tests

* Fix random state

natmokval committed:
TST: add the parameter MonthEnd(2) to 2 tests with freq= 2M (#53916)

TST: add the parameter MonthEnd(2) to some tests with freq= 2M

phofl committed:
ENH / CoW: Add lazy copy to eval (#53746)

jorisvandenbossche committed:
DOC: remove unnecessary note about EA values_for_factorize from 2.0.3 whatsnew (#53923)

DeaMariaLeon committed:
DOC: Changed M to MS in examples (#53927)

Changed M to MS in examples

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53909)

* Examples Resampler.asfreq, count, nunique, max, mean, median

* Example Resampler.min

* Added Returns section to min and max

* Changed M to MS and examples results accordingly

Charlie-XIAO committed:
DOC: EX01 (`Index` and `RangeIndex`) (#53920)

* Example for pd.Index.asof_locs

* Example for pd.Index.get_slice_bound

* Example for pd.RangeIndex

* Example for pd.RangeIndex.start/stop/step

* Example for pd.RangeIndex.from_range

* cleanup for now

* fix doctest

* example for RangeIndex.from_range is wrong

* retrigger checks

* retrigger checks

* retrigger checks

Charlie-XIAO committed:
CI: linting check to ensure `lib.NoDefault` is only used for typing (#53901)

* CI add liniting to check NoDefault used only for typing

* minor modification

* reduce cost

* tests added for the new linting check

* types_or -> types because only python

* rephrase pre-commit hook name

* rephrase more

* fix failing tests:

* retrigger checks

* retrigger checks

jbrockmendel committed:
REF: separate Block.pad_or_backfill from interpolate (#53918)

* REF: separate Block.pad_or_backfill from interpolate

* fix interp test

* mypy fixup

topper-123 committed:
BUG: fix Series.apply(..., by_row), v2. (#53601)

mroeschke committed:
CI: Bump 32 bit and musl linux jobs to PY3.11 (#53937)

jbrockmendel committed:
CLN: assorted (#53932)

Charlie-XIAO committed:
DOC: EX01 ({Categorical, Interval, Multi, Datetime, Timedelta}-Index) (#53925)

* add examples for CategoricalIndex.equals

* add examples for MultiIndex.dtypes

* add examples for MultiIndex.drop

* add examples for DatetimeIndex.to_pydatetime/std

* add examples for TimedeltaIndex

* minor modifications

* fix docstrings and make up for removed ones

* fix doctests

natmokval committed:
TST: add test with freq=2M to increase coverage of to_period (#53940)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53930)

* Examples Resampler.ohlc, prod, size, sem, std, sum, var

* Updated code_checks.sh

* Corrected sum docstring

* corrected default no, removed print

---------

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53946)

Examples options

mroeschke committed:
CI/TST: Misc test cleanups (#53931)

* TST: Misc testing cleanups

* more cleans

* pytest updates

* trigger ci

* Remove tmp_file removal

* Revert gcs testing

* type ignore

PrathumP committed:
CLN : Removed unnecessary numba-compat code (#53947)

* Removed unnecessary numba-compat code

* minor change

jbrockmendel committed:
REF: remove axis keyword from PandasArray.pad_or_backfill (#53941)

* REF: remove axis keyword from PandasArray.pad_or_backfill

* transpose

mroeschke committed:
TST: Always close figures in plotting tests (#53929)

* centeralize mpl_cleanup, make hist_df smaller

* remove def close

* trigger ci

* Bump to 50

rsm-23 committed:
TST: Concatenating categorical datetime columns (#53949)

Added unit-test

jorisvandenbossche committed:
DOC/WEB: update Google Analytics id (#53954)

MarcoGorelli committed:
DOC: fix pandas-coverage link (#53971)

fix pandas-coverage link

jbrockmendel committed:
PERF: ffill/bfill with non-numpy dtypes (#53950)

phofl committed:
ENH: Don't fragment manager if convert is no-op (#53977)

jorisvandenbossche committed:
DEPR: start with Deprecation instead of FutureWarning for NDFrame._data (#53994)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53948)

* Examples Rolling.max, cov, skew, apply

* correct skew

* Changing format skew

* Trying to fix skew

* Removed replace()

* Remove replace() correct doc.py links

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53985)

* Expanding examples

* change list for examples max, min

* Remove replace() on Expanding examples

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#53982)

* Examples Window.mean, sum, var, std

* Corrections to docstrings

* Removed replace()

* Correct link take 2

* remove replace() to right method

jbrockmendel committed:
DEPR: GroupBy.quantile with bool dtype (#53975)

rsm-23 committed:
DOC: Added note for `corr`  (#53972)

* Added note for corr

* Removed additional linebreak

* Added example and note for Series.corr

* Update frame.py

* Added operation in example

* Fixed indentation error for note

* Relocated notes

phofl committed:
Test CoW for multiple Python versions (#53981)

sqali committed:
Update ecosystem.md (#53980)

* Update ecosystem.md

replaced qgrid with modin-spreadsheet in docs

* Update ecosystem.md

removed the trailing spaces

phofl committed:
CoW: Add warning for chained assignment with fillna (#53779)

DeaMariaLeon committed:
DOC: Fixing EX01 - Added examples (#54004)

Added examples ewm

liang3zy22 committed:
BUG: Fix mamba can't create environment issue (#54017)

rhshadrach committed:
CLN: tests.groupby.test_any_all (#53998)

jbrockmendel committed:
BUG: missing fstring (#53990)

jbrockmendel committed:
BUG: ignoring sort in DTA.factorize (#53992)

jbrockmendel committed:
CLN: remove unreachable, unnecessary axis kwd (#53991)

wjandrea committed:
DOC: Clean up CSV sniffing and chunking examples (#53987)

jbrockmendel committed:
REF: helper for merge casting (#53976)

mroeschke committed:
CLN: Consolidate Dependencies (#53863)

* CLN: Dependencies

* Move back togglepromt

* Add tokenize-rt

* Use hauntsaninja/black-pre-commit-mirror

phofl committed:
CLN: Remove unnecessary pa version check (#54012)

rsm-23 committed:
DOC: Added to 10mins guide (#54010)

added basic intro

jbrockmendel committed:
CLN: is_mixed_type, is_homogeneous_type (#54008)

CLN: is_mixed_type

jbrockmendel committed:
REF: swap axis before calling Manager.pad_or_backfill (#53989)

rsm-23 committed:
DOC: Updated pandas extension list (#53960)

* Updated extension list

* Fixed typo

Charlie-XIAO committed:
API: add `NaTType` and `NAType` to `pandas.api.typing` (#53958)

* API: add NaTType and NAType to pandas.api.typing

* updated test

jbrockmendel committed:
DEPR: downcast keyword in Index.fillna (#53956)

* DEPR: downcast keyword in Index.fillna

* GH ref

* docstring fixup

phofl committed:
ENH: Use explicit methods instead of regex pattern in arrow strings (#54006)

* ENH: Use explicit methods instead of regex pattern in arrow strings

* Fixup

* Fix

rhshadrach committed:
DEPR: Replacing builtin and NumPy funcs in agg/apply/transform (#53974)

* DEPR: Replacing builtin and NumPy funcs in agg/apply/transform

* mypy fixup

natmokval committed:
TST: add test to check dtype after replacing values in categorical Series inplace (#53993)

* TST: add test to check dtype after replacing categorical Series inplace

* replace tm.assert_index_equal with tm.assert_series_equal

mroeschke committed:
TST: Mark subprocess tests as single_cpu (#54026)

penelopeysm committed:
DOC: Fix typos in groupby user guide (#54041)

Fix typos in groupby user guide

wcgonzal committed:
DOC: add backticks to docstrings (#54030)

add backticks to docstrings

mroeschke committed:
CI: Clean GH Action caches weekly (#53753)

* CI: Clean GH Action caches older than 1 week

* purge weekly instead

* ensure we list a lot of caches

harahu committed:
BUG: Fix failing hash funciton for certain non-ns resolution `Timedelta`s (#54037)

* Add reproducing test

* Attempt fix

* Add GH reference

* Add whatsnew entry

* Pre-commit after the fact

MilesCranmer committed:
BUG: Fix pandas compatibility with Python installations lacking bzip2 headers (#53858)

* BUG: Make bz2 import optional

* CLN: Create `get_bz2_file` to match `get_lzma_file`

* DOC: Add bz2 bugfix to changelog

* TST: Test bz2 non-import works

* TST: Test bz2 non-import from subprocess

* TST: Fix bz2 non-import test

* TST: Fix indentation issues in bz2 import test

* MAINT: Clean up merge commit

* Mark bz2 missing test with `single_cpu`

Daquisu committed:
Set copy=false after new data array, and adjust texts

jbrockmendel commented:
this was close to ready, re-opening.  @Daquisu can you merge main and address any outstanding comments

jbrockmendel commented:
@Daquisu can you merge main and address any outstanding comments?  ill make reviewing this a priority

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

Daquisu commented:
@jbrockmendel sorry for the _extremely_ long delay. Can you please review https://github.com/pandas-dev/pandas/pull/66537 when you get a chance? I helped @Flute-Gamer with the PR. Any high level guidance will help
